File: ui/timeline-panel/src/models/completionRate.ts
import type { TaskSlot } from "../types";
import {
  dateKey,
  filterTasksByCategory,
  type CategoryFilter,
} from "./taskStore";

export type CompletionTier = "zero" | "low" | "medium" | "high";

export type CompletionDayEntry = {
  rate: number;
};

/** Dates with labeled tasks only; absent keys mean no marker. */
export type CompletionRateMap = Record<string, CompletionDayEntry>;

export function calculateCompletionRate(tasks: TaskSlot[]): number | null {
  const activeTasks = tasks.filter((task) => task.label.trim() !== "");
  if (activeTasks.length === 0) {
    return null;
  }

  const completedCount = activeTasks.filter((task) => task.completed).length;
  return Math.round((completedCount / activeTasks.length) * 100);
}

export function getCompletionTier(rate: number): CompletionTier {
  if (rate === 0) {
    return "zero";
  }
  if (rate <= 30) {
    return "low";
  }
  if (rate <= 70) {
    return "medium";
  }
  return "high";
}

export function getCompletionTierLabel(rate: number): string {
  switch (getCompletionTier(rate)) {
    case "zero":
      return "미완료";
    case "low":
      return "낮음";
    case "medium":
      return "보통";
    case "high":
      return "높음";
  }
}

export function buildCompletionRateMap(
  store: Record<string, TaskSlot[]>,
  category: CategoryFilter = "전체",
): CompletionRateMap {
  const rates: CompletionRateMap = {};

  for (const [key, tasks] of Object.entries(store)) {
    const scopedTasks =
      category === "전체" ? tasks : filterTasksByCategory(tasks, category);
    const rate = calculateCompletionRate(scopedTasks);
    if (rate !== null) {
      rates[key] = { rate };
    }
  }

  return rates;
}

export function getCompletionEntryForDate(
  rates: CompletionRateMap,
  date: Date,
): CompletionDayEntry | undefined {
  return rates[dateKey(date)];
}

export function hasCompletionMarker(
  rates: CompletionRateMap,
  date: Date,
): boolean {
  return dateKey(date) in rates;
}

export function buildCalendarDateAriaLabel(
  month: number,
  day: number,
  entry: CompletionDayEntry | undefined,
): string {
  const base = `${month + 1}월 ${day}일 선택`;

  if (!entry) {
    return `${base}, 라벨 있는 일과 없음`;
  }

  const tierLabel = getCompletionTierLabel(entry.rate);
  return `${base}, 완료율 ${entry.rate}% (${tierLabel})`;
}

File: flutter/lib/models/completion_rate.dart
import 'category_store.dart';
import 'task_slot.dart';
import '../repositories/task_repository.dart';
import '../utils/date_key.dart';

enum CompletionTier { zero, low, medium, high }

class CompletionDayEntry {
  const CompletionDayEntry({required this.rate});

  final int rate;
}

typedef CompletionRateMap = Map<String, CompletionDayEntry>;

/// 웹 MVP `completionRate.ts`와 동일한 완료율 계산·티어 로직.
int? calculateCompletionRate(List<TaskSlot> tasks) {
  final activeTasks =
      tasks.where((task) => task.label.trim().isNotEmpty).toList();
  if (activeTasks.isEmpty) {
    return null;
  }

  final completedCount =
      activeTasks.where((task) => task.completed).length;
  return ((completedCount / activeTasks.length) * 100).round();
}

CompletionTier getCompletionTier(int rate) {
  if (rate == 0) {
    return CompletionTier.zero;
  }
  if (rate <= 30) {
    return CompletionTier.low;
  }
  if (rate <= 70) {
    return CompletionTier.medium;
  }
  return CompletionTier.high;
}

String getCompletionTierLabel(int rate) {
  switch (getCompletionTier(rate)) {
    case CompletionTier.zero:
      return '미완료';
    case CompletionTier.low:
      return '낮음';
    case CompletionTier.medium:
      return '보통';
    case CompletionTier.high:
      return '높음';
  }
}

CompletionRateMap buildCompletionRateMap(
  TaskStoreData store, [
  String category = CategoryStore.filterAll,
]) {
  final rates = <String, CompletionDayEntry>{};

  for (final entry in store.entries) {
    final scopedTasks = category == CategoryStore.filterAll
        ? entry.value
        : CategoryStore.filterTasksByCategory(entry.value, category);
    final rate = calculateCompletionRate(scopedTasks);
    if (rate != null) {
      rates[entry.key] = CompletionDayEntry(rate: rate);
    }
  }

  return rates;
}

CompletionDayEntry? getCompletionEntryForDate(
  CompletionRateMap rates,
  DateTime date,
) {
  return rates[dateKey(date)];
}

bool hasCompletionMarker(CompletionRateMap rates, DateTime date) {
  return rates.containsKey(dateKey(date));
}

String buildCalendarDateAriaLabel(
  DateTime date,
  CompletionDayEntry? entry,
) {
  final base = '${date.month}월 ${date.day}일 선택';

  if (entry == null) {
    return '$base, 라벨 있는 일과 없음';
  }

  final tierLabel = getCompletionTierLabel(entry.rate);
  return '$base, 완료율 ${entry.rate}% ($tierLabel)';
}
